src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ProjectRuleSnapshotExtensions.cs

```diff
@@ -67,5 +68,18 @@ public static IProjectRuleSnapshot GetSnapshotOrEmpty(this IImmutableDictionary<
 
             return EmptyProjectRuleSnapshot.Instance;
         }
+
+        public static bool HasChange(this IImmutableDictionary<string, IProjectChangeDescription> changesByRule)
+        {
+            foreach ((_, IProjectChangeDescription change) in changesByRule)
+            {
+                if (change.Difference.AnyChanges)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
```

Comment:
.Values allocates, this has to do with reducing allocates, these paths are very hot.

Reply:
Surprising. I figured both `.Keys` and `.Values` were more like a `Span` on the dictionary; in the sense that it simply has a the dictionary data and 'filters' what is being enumerated. If that's the case, then don't use `.Values`: `changesByRule.Any(kv => kv.Value.Difference.AnyChanges)` What's the downside of that, @davkean?

Reply:
Given the delegate will be cached due to it capturing no state, this will be idectical in allocations.

Reply:
The `foreach` avoids allocations when the collection is a concrete type rather than an interface. We could port the optimistic stuff from CPS to here to try and avoid the enumerator allocation.

Reply:
Huh. So, that enumerator allocation would apply to all LINQ methods, yes? Also, does your use of `ValueTuple<T>` incur allocations, or would it be better to `var`/`KeyValuePair<T>` the current element variable? Also, if CPS has good optimistic methods, we could see if they would want to expose them, or maybe have them as a separate package. Better to share code than duplicate it.

Reply:
`ValueTuple<T>` does not incur allocations, `ValueTuple<T>` is a struct.